### PR TITLE
Filesystem everywhere

### PR DESCRIPTION
**Pull request by halgari**
Rework how we use IFileSystem and AbsolutePaths. Due to the way several components work I don't think it's possible to keep both IFileSystem and Absolute paths separate, instead those places where we convert strings to AbsolutePaths are now *required* to supply an IFileSystem. I explained this in the attached ADR. 

I also removed KnownFolders, reworking all the code to go through `IFileSystem.GetKnownFolder`

### CODE REVIEW

File: src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs [hunk 1]
```diff
@@ -222,6 +222,15 @@ public interface IFileSystem
     /// <returns></returns>
     Task WriteAllTextAsync(AbsolutePath path, string text, CancellationToken cancellationToken = default);
 
+    /// <summary>
+    /// Writes all text to a file using UTF-8 encoding.
+    /// </summary>
+    /// <param name="path">Path to the file.</param>
+    /// <param name="text">Text to write to the file.</param>
+    /// <param name="cancellationToken">Optional <see cref="CancellationToken"/>.</param>
```
Comment (erri120) [inline src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs:230]: Should be removed.

File: tests/NexusMods.Paths.Tests/PathExpansionTests.cs [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
+    }
+    
     [Fact]
     public void CanExpandEntryFolder()
     {
         var inputPath = "{EntryFolder}/FluffyKitten.png";
-        var expandedPath = KnownFolders.ExpandPath(inputPath);
-        var expectedPath = Path.Combine(KnownFolders.EntryFolder.GetFullPath(), "FluffyKitten.png");
+        var expandedPath = _fileSystem.ExpandKnownFoldersPath(inputPath);
+        var expectedPath = Path.Combine(_fileSystem.GetKnownPath(KnownPath.EntryDirectory).GetFullPath(), "FluffyKitten.png");
```
Comment (erri120) [inline tests/NexusMods.Paths.Tests/PathExpansionTests.cs:20]: Tests like these should use `[Theory, AutoFileSystem]` using AutoFixture:

```csharp
[Theory, AutoFileSystem]
public void CanExpandEntryFolder(IFileSystem fs)

// or

[Theory, AutoFileSystem]
public void CanExpandEntryFolder(InMemoryFileSystem fs)
```

**Review by erri120 — CHANGES_REQUESTED** [review]
If we want to keep `IFileSystem` as a field of `AbsolutePath`, _every_ method of `IFileSystem` that operates on a single path, like `OpenFile`, should also have a method in `AbsolutePath` which redirects to `IFileSystem`:

```csharp
interface IFileSystem
{
    //...
    Stream WriteFile(AbsolutePath path);
}

struct AbsolutePath
{
    //...
    public Stream WriteFile() => FileSystem.WriteFile(this);
}
```

**Comment by erri120** [comment]
ARD `0005` is also already taken by #241.

File: src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs (hunk 1 above)
Comment (halgari) [reply src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs:230]: Why?

Comment (halgari) [reply src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs:230]: Without some amount of sync files we cause deadlocks in the some small situations where we need to not bounce to another thread while generating a file. Removing this and making `EnsureFiles` use `.Wait()` in "StubbedGame" causes a deadlock on systems (like CI) where we have 2 cores.

Comment (erri120) [reply src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs:230]: This was regarding the `cacnellationToken` parameter, which is unused for the sync methods. GitHub doesn't like it when someone just selects 1 line and shows multiple, so the context for this was likely weird.

File: tests/NexusMods.Paths.Tests/PathExpansionTests.cs (hunk 2 above)
Comment (halgari) [reply tests/NexusMods.Paths.Tests/PathExpansionTests.cs]: Updated to use Theory and InlineData, good catch

File: src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs (hunk 1 above)
Comment (halgari) [reply src/NexusMods.Paths/FileSystemAbstraction/IFileSystem.cs:230]: Ah, makes sense 😆

**Comment by halgari** [comment]
> If we want to keep `IFileSystem` as a field of `AbsolutePath`, _every_ method of `IFileSystem` that operates on a single path, like `OpenFile`, should also have a method in `AbsolutePath` which redirects to `IFileSystem`:
> 
> ```cs
> interface IFileSystem
> {
>     //...
>     Stream WriteFile(AbsolutePath path);
> }
> 
> struct AbsolutePath
> {
>     //...
>     public Stream WriteFile() => FileSystem.WriteFile(this);
> }
> ```

This may be the case in the long run, and we may want to move some helper methods out of FileSystem and into AbsolutePath_IO.cs. That may keep BaseFilesystem a bit cleaner

**Review by erri120 — APPROVED** [review]

**Pull request merged**